### ISSUE TITLE
`dart compile X` report errors inconsistently

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/sdk#54728]
The `dart compile js` reports errors on `stdout`:
```
% echo 'baz' > invalid.dart
% dart compile js invalid.dart 2> /dev/null
invalid.dart:1:1:
Error: Variables must be declared using the keywords 'const', 'final', 'var' or a type name.
baz
^^^
invalid.dart:1:1:
Error: Expected ';' after this.
baz
^^^
Error: No 'main' method found.
Error: Compilation failed.
```

The `dart compile exe` reports errors on `stderr`:
```
% echo 'baz' > invalid.dart
% dart compile exe invalid.dart 2> /dev/null
<silent>
% 
```

The test runner also expects e.g. for dart2js compilations errors to be produced on `stdout`.

=> I think all the `dart compile X` should report their errors on `stderr` as is common for normal unix tools.

/cc @bkonyi @mit-mit mentioned you may be the one to look at this.

**Comment by mit-mit** [comment]
Ben, it might be good to take a look at error reporting across not just the compile sub-command but rather all sub-commands?

**Comment by bkonyi** [comment]
I think this might actually be a dart2js issue (cc @sigmundch). `dart compile js` simply invokes the `dart2js` snapshot, so all the error output is printed there and dartdev has no control over which stream is being used. 

> Ben, it might be good to take a look at error reporting across not just the compile sub-command but rather all sub-commands?

I can reach out to the sub-command owners to do an audit. However, making any changes might be easier said than done. If there's any tooling relying on scraping stdout/stderr for a particular message from a command, it's going to break the moment we change it.

One extreme example is the VM service message. It's printed on stdout, which causes some headaches when trying to debug programs that use stdio to communicate since the message needs to be stripped from the beginning of the stream to not break the protocol. It really should be output on stderr. However, changing the message to not print on stdout would be a major breaking change and break almost every Dart and Flutter tool we support.

**Comment by mkustermann** [comment]
> If there's any tooling relying on scraping stdout/stderr for a particular message from a command, it's going to break the moment we change it.

Certainly our own tooling (namely `tools/test.py`) is expecting errors for some tools on stdout some on stderr. On dart2wasm I've made a recent change to emit & expect it on stderr (see [cl/347902](https://dart-review.googlesource.com/c/sdk/+/347902/8/pkg/test_runner/lib/src/command_output.dart))

> As far as dart2js goes, I'm open to push for the breaking change. I filed #54733 to track that.

Is this really breaking? I'd assume we maintain the invariant that if there's errors there's a non-0 exit code and if there's no errors `stderr` is empty and exit code is 0.

**Comment by sigmundch** [comment]
> Is this really breaking? I'd assume we maintain the invariant that if there's errors there's a non-0 exit code and if there's no errors stderr is empty and exit code is 0.

Maybe it's not an official break, but if it may affect people, I rather over communicate ahead of time. I wouldn't be surprised if some tools are already expecting the old behavior to properly plumb data. As an example, I haven't looked yet, but I can imagine build_runner and flutter-tools may have some assumptions to forward errors from stdout at this time because that was the existing behavior. I hope they already forward stderr because they reuse logic across processes invocations, but worth checking nonetheless.

**Comment by bkonyi** [comment]
Working on an unrelated change, I've noticed that the behavior for `dart compile exe` with warnings is different depending on whether or not there's an error present in the program.

For example, consider the following Dart program:

```dart
// foo.dart
void main() {
  int i = 0;
  i?.isEven;
}
```

If I run `dart compile exe foo.dart`, a warning is printed to `stdout`:

```
Compiling foo.dart to kernel file foo.exe.
foo.dart:3:3: Warning: Operand of null-aware operation '?.' has type 'int' which excludes null.
  i?.isEven;
  ^
```

However, if I remove a semicolon to introduce an error, both the error and the warning are output on `stderr`:

```
// stdout start
Compiling foo.dart to kernel file foo.exe.

// stderr start
foo.dart:2:11: Error: Expected ';' after this.
  int i = 0
          ^
foo.dart:3:3: Warning: Operand of null-aware operation '?.' has type 'int' which excludes null.
  i?.isEven;
  ^
```

It seems strange that warnings would be printed on `stdout` in one case and `stderr` in the other. What do you think the correct behavior should be here @mkustermann?

**Comment by mkustermann** [comment]
> It seems strange that warnings would be printed on stdout in one case and stderr in the other. What do you think the correct behavior should be here @mkustermann?

Seems the code does it intentionally in https://github.com/dart-lang/sdk/blob/4dff1b3b0ad72d64ccd0/pkg/dart2native/lib/generate.dart#L132

IMHO if a tool exits with non-zero status, I'd expect the `stderr` to contain the reason for the non-zero exit code (there may be tons of verbose stuff on `stdout` that I can filter away to `/dev/null` and I'm left with the actual errors that made things go wrong). So for info/warnings that don't cause compilation failure, I'd make them go to `stdout`.

But whatever we settle on, it makes sense to have consistency.

**Comment by mkustermann** [comment]
> I agree in general. However, this isn't ideal for the VM. Running the above program will result in the warning message being output to stderr, not stdout. I think the VM needs to be a special case since polluting stdout when running user code could break workflows relying on scraping stdout.

Fair point about not compiling but running from source. Though there's an alternative solution as well: Suppress info/warnings when running from source entirely (possibly emit in `--verbose` or so mode). (If I run a Python program that uses calls a function that's marked deprecated, would it issue any info/warning to stdout or stderr?)
